loader/loader.c

```diff
@@ -5912,52 +5913,102 @@ VkResult setup_loader_term_phys_devs(struct loader_instance *inst) {
         if (VK_SUCCESS != res) {
             goto out;
         }
-        inst->total_gpu_count += icd_phys_dev_array[icd_idx].device_count;
         icd_phys_dev_array[icd_idx].icd_term = icd_term;
         icd_phys_dev_array[icd_idx].icd_index = icd_idx;
         icd_term = icd_term->next;
         ++icd_idx;
     }
 
-    if (0 == inst->total_gpu_count) {
+    // Add up both the windows sorted and non windows found physical device counts
+    for (uint32_t i = 0; i < windows_sorted_devices_count; ++i) {
+        new_phys_devs_count += windows_sorted_devices_array[i].device_count;
+    }
+    for (uint32_t i = 0; i < icd_phys_dev_count; ++i) {
+        new_phys_devs_count += icd_phys_dev_array[i].device_count;
+    }
+
+    // Bail out if there are no physical devices reported
+    if (0 == new_phys_devs_count) {
         loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
                    "setup_loader_term_phys_devs:  Failed to detect any valid GPUs in the current config");
         res = VK_ERROR_INITIALIZATION_FAILED;
         goto out;
     }
 
-    new_phys_devs = loader_instance_heap_alloc(inst, sizeof(struct loader_physical_device_term *) * inst->total_gpu_count,
+    // Create an allocation large enough to hold both the windows sorting enumeration and non-windows physical device enumeration
+    new_phys_devs = loader_instance_heap_alloc(inst, sizeof(struct loader_physical_device_term *) * new_phys_devs_count,
                                                VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
     if (NULL == new_phys_devs) {
         loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
-                   "setup_loader_term_phys_devs:  Failed to allocate new physical device array of size %d", inst->total_gpu_count);
+                   "setup_loader_term_phys_devs:  Failed to allocate new physical device array of size %d", new_phys_devs_count);
         res = VK_ERROR_OUT_OF_HOST_MEMORY;
         goto out;
     }
-    memset(new_phys_devs, 0, sizeof(struct loader_physical_device_term *) * inst->total_gpu_count);
+    memset(new_phys_devs, 0, sizeof(struct loader_physical_device_term *) * new_phys_devs_count);
+
+    // Current index into the new_phys_devs array - increment whenever we've written in.
+    uint32_t idx = 0;
+
+    // Copy over everything found through sorted enumeration
```

Comment:
Wouldn't it be better, and more obvious to put this in an else associated with the #ifdef on line 5989?

Reply:
No because it is the 'currently written index in `new_phys_devs`. 
The logic is that we want to copy over all the sorted physical devices gotten through the windows specific mechanism, then copy in the rest of the phys devs from the non-windows sorting path.

Reply:
But this is "windows" sorting logic where the if below is "linux" sorting logic.  They don't ever run together do they?